src/XPurchase/VirtoCommerce.XPurchase/Services/CartProductService.cs

```diff
@@ -58,23 +48,40 @@ public async Task<IEnumerable<CartProduct>> GetCartProductsByIdsAsync(CartAggreg
                 });
 
                 var pricesEvalContext = _mapper.Map<PriceEvaluationContext>(cartAggr);
-                pricesEvalContext.ProductIds = ids;             
+                pricesEvalContext.ProductIds = ids;
                 var evalPricesTask = _pricingService.EvaluateProductPricesAsync(pricesEvalContext);
 
-                await Task.WhenAll(loadInventoriesTask, evalPricesTask);
-              
-                foreach (var product in products)
+                await Task.WhenAll(loadInventoriesTask, evalPricesTask);
+
+                foreach (var cartProduct in result)
                 {
-                    var cartProduct = new CartProduct(product);
                     //Apply inventories
                     cartProduct.ApplyInventories(loadInventoriesTask.Result.Results, cartAggr.Store);
 
                     //Apply prices
                     cartProduct.ApplyPrices(evalPricesTask.Result, cartAggr.Currency);
-                    result.Add(cartProduct);
                 }
             }
             return result;
+        }
+
+        public async Task<IEnumerable<CartProduct>> GetProductsByIdsAsync(CartAggregate cartAggr, string[] ids, string additionalResponseGroups = null)
+        {
+            if (cartAggr == null)
+            {
+                throw new ArgumentNullException(nameof(cartAggr));
+            }
+            if (ids == null)
+            {
+                throw new ArgumentNullException(nameof(ids));
+            }

```

Comment:
```suggestion
            var tcs = new TaskCompletionSource<IEnumerable<CartProduct>>();
           
            if (cartAggr == null)
            {
                tcs.SetException(new ArgumentNullException(nameof(cartAggr)));
                return tcs.Task;
            }

            if (ids == null)
            {
                tcs.SetException(new ArgumentNullException(nameof(ids)));
                return tcs.Task;
            }

```

Reply:
Error	CS4016	Since this is an async method, the return expression must be of type 'IEnumerable<CartProduct>' rather than 'Task<IEnumerable<CartProduct>>'